tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileSystemManagedLedgerOffloader.java

```diff
@@ -188,12 +191,15 @@ public void run() {
                 AtomicLong haveOffloadEntryNumber = new AtomicLong(0);
                 long needToOffloadFirstEntryNumber = 0;
                 CountDownLatch countDownLatch;
+                //avoid prefetch too much data into memory
+                ArrayBlockingQueue<Boolean> tasks = new ArrayBlockingQueue<>(PREFETCH_ROUNDS);
                 do {
                     long end = Math.min(needToOffloadFirstEntryNumber + ENTRIES_PER_READ - 1, readHandle.getLastAddConfirmed());
                     log.debug("read ledger entries. start: {}, end: {}", needToOffloadFirstEntryNumber, end);
                     LedgerEntries ledgerEntriesOnce = readHandle.readAsync(needToOffloadFirstEntryNumber, end).get();
```

Comment:
the blockingQueue is to block the whole do while loop(to be precisely, the read and submit) Here, not try to limit fileSystemWriter. I understand that the FileSystemWriter will be executed orderly. My intention is to limit the whole do while loop because the do while loop((readAsync().get()) read data from ledger in advance too much too quickly(and lead to consume much memory), more than  FileSystemWriter Threads can handle timely. So read in advance, or put in other words, prefetch, should be controlled to some degree. So I introduce a blocking queue and hence a producer-consumer mock-up. the do while loop(read and submit) acts like a producer, the FileSystemWriter acts as a consumer. @congbobo184

Reply:
I have  understood your min, I look up the code context, I think you can use Semaphore to fix it :)

Reply:
nice idea! I will change the code to implement the idea @congbobo184